ruby/ql/lib/codeql/ruby/security/CleartextLoggingCustomizations.qll

```diff
@@ -0,0 +1,311 @@
+ * adding your own.
+ */
+
+private import ruby
+private import codeql.ruby.DataFlow
+private import codeql.ruby.TaintTracking::TaintTracking
+private import codeql.ruby.Concepts
+private import codeql.ruby.dataflow.RemoteFlowSources
+private import internal.SensitiveDataHeuristics::HeuristicNames
+private import codeql.ruby.CFG
+private import codeql.ruby.dataflow.SSA
+
+/**
+ * Provides default sources, sinks and sanitizers for reasoning about
+ * cleartext logging of sensitive information, as well as extension points for
+ * adding your own.
+ */
+module CleartextLogging {
+  /**
+   * A data flow source for cleartext logging of sensitive information.
+   */
+  abstract class Source extends DataFlow::Node {
+    /** Gets a string that describes the type of this data flow source. */
+    abstract string describe();
+  }
+
+  /**
+   * A data flow sink for cleartext logging of sensitive information.
+   */
+  abstract class Sink extends DataFlow::Node { }
+
+  /**
+   * A sanitizer for cleartext logging of sensitive information.
+   */
+  abstract class Sanitizer extends DataFlow::Node { }
+
+  /**
+   * A node that receives sanitized sensitive information.
+   */
+  abstract class SanitizerIn extends DataFlow::Node { }
+
+  /**
+   * Holds if `re` may be a regular expression that can be used to sanitize
+   * sensitive data with a call to `sub`.
+   */
+  private predicate effectiveSubRegExp(RegExpLiteral re) {
+    re.getConstantValue().getStringOrSymbol().matches([".*", ".+"])
+  }
+
+  /**
+   * Holds if `re` may be a regular expression that can be used to sanitize
+   * sensitive data with a call to `gsub`.
+   */
+  private predicate effectiveGsubRegExp(RegExpLiteral re) {
+    re.getConstantValue().getStringOrSymbol().matches(".")
+  }
+
+  /**
+   * A call to `sub`/`sub!` or `gsub`/`gsub!` that seems to mask sensitive information.
+   */
+  private class MaskingReplacerSanitizer extends Sanitizer, DataFlow::CallNode {
+    MaskingReplacerSanitizer() {
+      exists(RegExpLiteral re |
+        re = this.getArgument(0).asExpr().getExpr() and
+        (
+          this.getMethodName() = ["sub", "sub!"] and effectiveSubRegExp(re)
+          or
+          this.getMethodName() = ["gsub", "gsub!"] and effectiveGsubRegExp(re)
+        )
+      )
+    }
+  }
+
+  /**
+   * A node sanitized by a prior call to `sub!` or `gsub!`,
+   * e.g. the `password` argument to `info` in:
+   * ```
+   * password = "changeme"
+   * password.sub!(/.+/, "")
+   * Logger.new(STDOUT).info password
+   * ```
+   */
+  private class MaskingReplacerSanitizedNode extends SanitizerIn {
+    MaskingReplacerSanitizedNode() {
+      exists(MaskingReplacerSanitizer maskCall, Variable v |
+        maskCall.getMethodName() = ["sub!", "gsub!"] and
+        v = maskCall.getReceiver().asExpr().getExpr().(VariableReadAccess).getVariable() and
+        v = this.asExpr().getExpr().(VariableReadAccess).getVariable() and
+        maskCall.asExpr().getASuccessor*() = this.asExpr()
```

Comment:
I think that limiting restricting this to `SanitizerIn` rather than just `Sanitizer` was unnecessary, I'll change this to just being a `Sanitizer` and remove the `SanitizerIn` related code in this query.

As a side note (and a reminder to myself), this is a separate case to `MaskingReplacerSanitizer` because `sub!`/`gsub!` mutate the receiver object. In the case of:
```
foo = "password"
foo = foo.sub(/.*/, "[redacted]")
logger.info(foo)
```
there is an edge from the second definition of `foo = foo.sub(/.*/, "[redacted]")` to the read of `foo` in the `logger.info(foo)` call and no edge directly from the first definition `foo = "password"` to the last read of `foo` because `foo` is redefined. On the other hand, in:
```
foo = "password"
foo.sub!(/.*/, "[redacted]")
logger.info(foo)
```
the edge is from the definition of `foo = "password"` straight to the read of `foo` in the `logger.info(foo)` call, without any edges involving the `sub!` call.